File: platforms/android/foundation/foundation/src/main/java/io/nativeblocks/foundation/modifier/Shadow.kt
package io.nativeblocks.foundation.modifier

import androidx.compose.ui.draw.shadow
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import io.nativeblocks.foundation.util.SHAPE_RECTANGLE
import io.nativeblocks.foundation.util.cornerShape
import io.nativeblocks.compiler.type.Modifier
import io.nativeblocks.compiler.type.ModifierData
import androidx.compose.ui.Modifier as UiModifier

@Modifier(
    keyType = "nativeblocks/shadow",
    name = "Shadow",
    description = "Draws an elevation shadow behind the block.",
    version = 1,
    versionName = "1",
)
internal fun shadow(
    @ModifierData(description = "Shadow elevation in DP.", defaultValue = "0.0")
    elevation: Dp = 0.dp,
    @ModifierData(description = "Shape of the shadow (rectangle, circle).", defaultValue = "rectangle")
    shape: String = SHAPE_RECTANGLE,
    @ModifierData(description = "Top-start corner radius in DP.", defaultValue = "0.0")
    radiusTopStart: Dp = 0.dp,
    @ModifierData(description = "Top-end corner radius in DP.", defaultValue = "0.0")
    radiusTopEnd: Dp = 0.dp,
    @ModifierData(description = "Bottom-start corner radius in DP.", defaultValue = "0.0")
    radiusBottomStart: Dp = 0.dp,
    @ModifierData(description = "Bottom-end corner radius in DP.", defaultValue = "0.0")
    radiusBottomEnd: Dp = 0.dp,
    @ModifierData(description = "Whether the content is also clipped to the shape.", defaultValue = "true")
    clipToShape: Boolean = true,
): UiModifier = UiModifier.shadow(
    elevation = elevation,
    shape = cornerShape(shape, radiusTopStart, radiusTopEnd, radiusBottomStart, radiusBottomEnd),
    clip = clipToShape,
)

File: platforms/ios/foundation/Sources/NativeblocksFoundation/modifier/Shadow.swift
import NativeblocksCompiler
import NativeblocksRuntime
import SwiftUI

@Modifier(
    name: "Shadow",
    keyType: "nativeblocks/shadow",
    description: "Drops a shadow behind the block.",
    version: 1,
    versionName: "1"
)
struct Shadow: ViewModifier {

    @ModifierData(description: "Shadow blur radius.", defaultValue: "0.0")
    var radius: CGFloat = 0

    @ModifierData(description: "Shape of the shadow (rectangle, circle).", defaultValue: "rectangle")
    var shape: String = SHAPE_RECTANGLE

    @ModifierData(description: "Top-leading corner radius.", defaultValue: "0.0")
    var radiusTopLeading: CGFloat = 0

    @ModifierData(description: "Top-trailing corner radius.", defaultValue: "0.0")
    var radiusTopTrailing: CGFloat = 0

    @ModifierData(description: "Bottom-leading corner radius.", defaultValue: "0.0")
    var radiusBottomLeading: CGFloat = 0

    @ModifierData(description: "Bottom-trailing corner radius.", defaultValue: "0.0")
    var radiusBottomTrailing: CGFloat = 0

    @ModifierData(description: "Whether the content is also clipped to the shape.", defaultValue: "true")
    var clipToShape: Bool = true

    @Environment(\.layoutDirection) var layoutDirection

    @ViewBuilder
    func body(content: Content) -> some View {
        if clipToShape {
            content
                .clipShape(
                    CornerShape(
                        shape: shape,
                        topLeading: radiusTopLeading,
                        topTrailing: radiusTopTrailing,
                        bottomLeading: radiusBottomLeading,
                        bottomTrailing: radiusBottomTrailing,
                        layoutDirection: layoutDirection
                    )
                )
                .shadow(radius: radius, y: radius / 2)
        } else {
            content.shadow(radius: radius, y: radius / 2)
        }
    }
}
